Firestore/CMakeLists.txt

```diff
@@ -15,6 +15,53 @@
 cmake_minimum_required(VERSION 2.8.11)
 project(firestore C CXX)
 
+option(ASAN "Build with Address sanitizer (mutually exculsive with TSAN)" OFF)
+# TODO(varconst): msan
+# Memory sanitizer is more complicated:
+# - it requires all dependencies to be compiled with msan enabled (see
+#   https://github.com/google/sanitizers/wiki/MemorySanitizerLibcxxHowTo);
+# - AppleClang doesn't support it.
+option(TSAN "Build with Thread sanitizer (mutually exculsive with ASAN)" OFF)
+option(UBSAN "Build with Undefined behavior sanitizer" OFF)
+
+if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  set(HAVE_SANITIZERS 1)
+endif()
+if(ASAN OR TSAN OR UBSAN)
+  set(USE_SANITIZERS 1)
+endif()
+
+if (NOT HAVE_SANITIZERS AND USE_SANITIZERS)
+  message(FATAL_ERROR "Only Clang and GCC support sanitizers.")
+endif()
+
+function(add_to_compile_and_link_flags flag)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${flag}" PARENT_SCOPE)
+  set(CMAKE_LINKER_FLAGS "${CMAKE_LINKER_FLAGS} ${flag}" PARENT_SCOPE)
+endfunction()
+
+if (HAVE_SANITIZERS AND USE_SANITIZERS)
+  if ((ASAN AND TSAN) OR (UBSAN AND TSAN))
+    message(FATAL_ERROR "You cannot combine sanitizers except asan and ubsan.")
+  endif()
+
+  # Recommended to "get nicer stack traces in error messages"
+  # TODO(varconst): double-check that tsan actually needs this flag (it's
+  # explicitly recommended in the docs for asan and ubsan)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-omit-frame-pointer")
+
+  if (TSAN)
+    add_to_compile_and_link_flags("-fsanitize=thread")
```

Comment:
Doesn't tsan also need us to specify `-pie` on final linking and `-fPIC` during compilation?

Reply:
Hmm, apparently not. The most relevant thing I could find is [this thread](https://github.com/arsenm/sanitizers-cmake/issues/13), which doesn't contain a definitive answer, but has [this quote](https://clang.llvm.org/docs/ThreadSanitizer.html#limitations) from the Clang docs:

> Non-position-independent executables are not supported. Therefore, the `fsanitize=thread` flag will cause Clang to act as though the `-fPIE` flag had been supplied if compiling without `-fPIC`, and as though the `-pie` flag had been supplied if linking an executable.

I couldn't find the same stated explicitly for GCC. However, the wiki page for TSan only has [the following instructions](https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual#usage):

> Simply compile your program with -fsanitize=thread and link it with -fsanitize=thread. To get a reasonable performance add -O2. Use -g to get file names and line numbers in the warning messages.

(Let me know if you think `-g` should be added)